File: diploma-server/server-data/src/test/java/org/arhor/diploma/data/persistence/repository/DatabaseIntegrationTest.java
package org.arhor.diploma.data.persistence.repository;

import org.arhor.diploma.TestConfig;
import org.arhor.diploma.testutils.IntegrationTest;
import org.arhor.diploma.testutils.RandomParameter;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.junit.jupiter.SpringExtension;
import org.testcontainers.containers.PostgreSQLContainer;
import org.testcontainers.junit.jupiter.Container;
import org.testcontainers.junit.jupiter.Testcontainers;

@IntegrationTest
@SpringBootTest(classes = {TestConfig.class})
@Testcontainers(disabledWithoutDocker = true)
@ExtendWith({RandomParameter.Resolver.class, SpringExtension.class})
public abstract class DatabaseIntegrationTest {

    @Container
    private static final PostgreSQLContainer<?> POSTGRES =
            new PostgreSQLContainer<>("postgres:11.7")
                    .withDatabaseName("diploma_test_db")
                    .withUsername("postgres")
                    .withPassword("password");
}

File: diploma-server/module-main/src/test/kotlin/org/arhor/diploma/data/persistence/repository/DatabaseIntegrationTest.kt
package org.arhor.diploma.data.persistence.repository;

import org.arhor.diploma.config.DatabaseConfig
import org.arhor.diploma.testutils.RandomParameter
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.boot.test.autoconfigure.data.r2dbc.DataR2dbcTest
import org.springframework.test.context.ContextConfiguration
import org.springframework.test.context.DynamicPropertyRegistry
import org.springframework.test.context.DynamicPropertySource
import org.testcontainers.containers.PostgreSQLContainer
import org.testcontainers.junit.jupiter.Container
import org.testcontainers.junit.jupiter.Testcontainers

@DataR2dbcTest
@Testcontainers(disabledWithoutDocker = true)
@ExtendWith(RandomParameter.Resolver::class)
@ContextConfiguration(classes = [DatabaseConfig::class])
abstract class DatabaseIntegrationTest {

    companion object {
        @JvmStatic
        @Container
        private val db = PostgreSQLContainer<Nothing>("postgres:11.7")

        @JvmStatic
        @DynamicPropertySource
        fun registerDynamicProps(registry: DynamicPropertyRegistry) {

            val databaseUrl = "postgresql://${db.host}:${db.firstMappedPort}/${db.databaseName}"
            val username = db.username
            val password = db.password

            registry.apply {
                add("spring.r2dbc.url") { "r2dbc:${databaseUrl}" }
                add("spring.r2dbc.username") { username }
                add("spring.r2dbc.password") { password }

                add("spring.flyway.url") { "jdbc:${databaseUrl}" }
                add("spring.flyway.user") { username }
                add("spring.flyway.password") { password }
                add("spring.flyway.locations") { "classpath:/db/migration" }
                add("spring.flyway.enabled") { true }
            }
        }
    }
}
